curvlinops/kfac.py

```diff
@@ -346,37 +434,22 @@ def _matmat(self, M: List[Tensor]) -> List[Tensor]:
             ):
                 w_pos, b_pos = param_pos["weight"], param_pos["bias"]
                 # v denotes the free dimension for treating multiple vectors in parallel
-                M_w = rearrange(M[w_pos], "c_out ... v -> c_out (...) v")
+                M_w = rearrange(M[w_pos], "m ... v -> m (...) v")
                 M_joint = cat([M_w, M[b_pos].unsqueeze(-2)], dim=-2)
-                aaT = self._input_covariances[mod_name]
-                ggT = self._gradient_covariances[mod_name]
-                M_joint = einsum(ggT, M_joint, aaT, "i j, j k v, k l -> i l v")
-
+                M_joint = self._left_and_right_multiply(M_joint, aaT, ggT)
                 w_cols = M_w.shape[1]
                 KM[w_pos], KM[b_pos] = M_joint.split([w_cols, 1], dim=-2)
                 KM[b_pos].squeeze_(1)
-
-            # for weights we need to multiply from the right with aaT
-            # for weights and biases we need to multiply from the left with ggT
             else:
-                for p_name, pos in param_pos.items():
-                    if p_name == "weight":
-                        M_w = rearrange(M[pos], "c_out ... v -> c_out (...) v")
-                        aaT = self._input_covariances[mod_name]
-                        KM[pos] = einsum(M_w, aaT, "c_out j v, j k -> c_out k v")
-                    else:
-                        KM[pos] = M[pos]
-
-                    ggT = self._gradient_covariances[mod_name]
-                    KM[pos] = einsum(ggT, KM[pos], "j k, k ... v -> j ... v")
+                self._separate_left_and_right_multiply(KM, M, param_pos, aaT, ggT)
 
             # restore original shapes
             if "weight" in param_pos:
                 KM[param_pos["weight"]] = KM[param_pos["weight"]].view(weight_shape)
 
         return KM
 
-    def _compute_kfac(self):
+    def compute_kronecker_factors(self):
```

Comment:
Is this method public on purpose now?

Reply:
Yes, I believe it makes sense to expose this to the user. For example, imagine the setting where you want to fit EKFAC in two steps: 1) first, you set `check_deterministic=False`, 2) you fit the Kronecker factors with `compute_kronecker_factors` and save to disk, 2) you load from disk in a different job and compute the eigenvalue corrections.